### DISCUSSION TITLE
Best practice way to detect Optional

### DISCUSSION THREAD

**Discussion by Floppy012** [spatie/laravel-data#315] [Q&A] [answered]
I'm having a DTO and need to perform an action on a `DataCollection` that can be optional.

Currently I check for `Optional` using `instanceof` like this:

```php
if (!($dto->collection instanceof Optional)) {
    // Do something
}
```

it kind of feels wrong to do it like this. Is there any better way to do this?

**Comment by rubenvanassche** [comment, marked as answer]
I think this is the best way to do this 😄

**Reply by ejunker** [reply]
@rubenvanassche what do you think about adding a `has()` method so you could do something like this:

```php
if ($dto->has('collection')) {
  // Do something
}
```

This is useful when you are using Data objects created from a Request and you are doing a partial field update. This is similar to `has()` method in `Illuminate\Http\Request`

**Reply by ejunker** [reply]
Maybe it would look something like this:
```php
    public function has(string $propertyName): bool
    {
        return isset($this->{$propertyName}) && ! ($this->{$propertyName} instanceof Optional);
    }
```

**Reply by onlime** [reply]
Thanks @ejunker for your solution. Still missing this in current Laravel-data. It just does not feel "clean" to use `Optional` object anywhere outside of the data objects.

My current trait looks like this:

```php
namespace App\Traits;

use Spatie\LaravelData\Data;
use Spatie\LaravelData\Optional;

/** @mixin Data */
trait HasDataAccessHelpers
{
    public function has(string $propertyName): bool
    {
        return isset($this->{$propertyName}) && ! ($this->{$propertyName} instanceof Optional);
    }

    public function get(string $propertyName, mixed $default = null): mixed
    {
        return $this->has($propertyName) ? $this->{$propertyName} : $default;
    }
}
```

Should we PR such helper methods or did you find a better way of handling this in the meantime? I'd prefer using `isset()`, overriding it with the `__isset()` in the data objects, but that would be quite obtrusive.

**Reply by drmmr763** [reply]
Coming here looking for this exact solution. Adding a `has` and `get` would be super useful. Until then borrowing @onlime 's trait!

**Reply by maxanstey-acquire** [reply]
I've started a new discussion around "ommitable" properties for awareness: https://github.com/spatie/laravel-data/discussions/984
